File: src/main/com/mega/game/engine/entities/contracts/IBodyEntity.java
package com.mega.game.engine.entities.contracts;

import com.badlogic.gdx.math.Vector2;
import com.mega.game.engine.common.ClassInstanceUtils;
import com.mega.game.engine.common.interfaces.IPositional;
import com.mega.game.engine.common.interfaces.ISizable;
import com.mega.game.engine.entities.IGameEntity;
import com.mega.game.engine.entities.IGameEntity;
import com.mega.game.engine.world.Body;
import com.mega.game.engine.world.BodyComponent;
import kotlin.reflect.KClass;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

/**
 * Interface for entities that have a body.
 */
public interface IBodyEntity extends IGameEntity, IPositional, ISizable {

    default Body getBody() {
        KClass<BodyComponent> key = ClassInstanceUtils.convertToKClass(BodyComponent.class);
        return Objects.requireNonNull(getComponent(key)).getBody();
    }

    default float getX() {
        return getBody().getX();
    }

    default void setX(float x) {
        getBody().setX(x);
    }

    default float getY() {
        return getBody().getY();
    }

    default void setY(float y) {
        getBody().setY(y);
    }

    default void setPosition(float x, float y) {
        getBody().setPosition(x, y);
    }

    @NotNull
    default Vector2 getPosition() {
        return getBody().getPosition();
    }

    default float getWidth() {
        return getBody().getWidth();
    }

    default void setWidth(float width) {
        getBody().setWidth(width);
    }

    default float getHeight() {
        return getBody().getHeight();
    }

    default void setHeight(float height) {
        getBody().setHeight(height);
    }

    default void setSize(float width, float height) {
        getBody().setSize(width, height);
    }

    default void translateSize(float width, float height) {
        getBody().setWidth(getBody().getWidth() + width);
        getBody().setHeight(getBody().getHeight() + height);
    }

    default void translate(float x, float y) {
        getBody().translation(x, y);
    }
}

File: src/main/com/mega/game/engine/entities/contracts/IBodyEntity.kt
package com.mega.game.engine.entities.contracts

import com.mega.game.engine.common.interfaces.IPositional
import com.mega.game.engine.common.interfaces.ISizable
import com.mega.game.engine.entities.IGameEntity
import com.mega.game.engine.world.body.Body
import com.mega.game.engine.world.body.BodyComponent

interface IBodyEntity : IGameEntity, IPositional, ISizable {

    val bodyComponent: BodyComponent
        get() {
            val key = BodyComponent::class
            return getComponent(key)!!
        }

    val body: Body
        get() = bodyComponent.body

    override fun getX(): Float {
        return this.body.getX()
    }

    override fun setX(x: Float) {
        this.body.setX(x)
    }

    override fun getY(): Float {
        return this.body.getY()
    }

    override fun setY(y: Float) {
        this.body.setY(y)
    }

    override fun setPosition(x: Float, y: Float) {
        this.body.setPosition(x, y)
    }

    override fun getWidth(): Float {
        return this.body.getWidth()
    }

    override fun setWidth(width: Float) {
        this.body.setWidth(width)
    }

    override fun getHeight(): Float {
        return this.body.getHeight()
    }

    override fun setHeight(height: Float) {
        this.body.setHeight(height)
    }

    override fun setSize(width: Float, height: Float) {
        this.body.setSize(width, height)
    }

    override fun translateSize(width: Float, height: Float) {
        this.body.setWidth(this.body.getWidth() + width)
        this.body.setHeight(this.body.getHeight() + height)
    }

    override fun translate(x: Float, y: Float) {
        this.body.translate(x, y)
    }
}